README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# The lg package for calculating local Gaussian correlations in multivariate applications

Otneim and Tjøstheim (2017) describes a new method for estimating
multivariate density functions using the concept of local Gaussian
correlations. Otneim and Tjøstheim (2018) expands the idea to the
estimation of *conditional* density functions. This package, written for
the R programming language, provides a simple interface for implementing
these methods in practical problems.

# Construction of the lg-object

Let us illustrate the use of this package by looking at the built-in
data set of daily closing prices of 4 major European stock indices in
the period 1991-1998. We load the data and transform them to daily
returns:

``` r
data(EuStockMarkets)
x <- apply(EuStockMarkets, 2, function(x) diff(log(x)))

# Remove the days where at least one index did not move at all
x <- x[!apply(x, 1, function(x) any(x == 0)),]
```

When using this package, the first task is always to create an lg-object
using the lg\_main()-function. This object contains all the estimation
parameters that will be used in the estimation step, including the
bandwidths. There are three main parameters that one can tune:

  - **The bandwidth selection method**: The bandwidth selection method
    is controlled through the argument *bw\_method*. It can take one of
    two values:
      - Use *bw\_method = “cv”* to use the cross-validation routine
        described by Otneim and Tjøstheim (2017). Depending on your
        system, this method is fairly slow and may take several minutes
        even for moderately sized data sets.
      - Use *bw\_method = “plugin”* for plugin bandwidths. This is the
        **default** method and very quick. It simply sets all bandwidths
        equal to 1.75\*n^(-1/6), which is derived from the asymptotic
        convergence rates for the local Gaussian correlations. Both
        numbers (1.75 and -1/6) can be set manually (see documentation
        of the lg\_main()-function).
  - **The estimation method:** The method of estimation is controlled
    through the argument *est\_method*. It can take one of three values:
      - Otneim and Tjøstheim (2017) uses a simplified method for
        multivariate density estimation. The density estimate is a
        locally Gaussian distribution, with correlations being estimated
        *locally* and *pairwise*. The data is transformed for marginal
        standard normality (see next point) and as a consequence, we fix
        the means and standard deviations to 0 and 1 respectively. To
        use this estimation method, write *est\_method = “1par”*. This
        is the **default** method.
      - Set *est\_method = “5par\_marginals\_fixed”* to estimate local
        means and local standard deviations marginally, as well as the
        pairwise local correlations. This is a more flexible method, but
        its theoretical properties are not (yet) fully understood. This
        configuration allows for the estimation of multivariate density
        functions without having to transform the data.
      - The option *est\_method = “5par”* is reserved to bivariate
        problems, and is a fully nonparametric estimation method as laid
        out by Tjøstheim & Hufthammer (2013). This will simply invoke
        the *localgauss* package (Berentsen et. al., ).
  - **Transformation of the marginals:** This is controlled by the
    logical argument *transform\_to\_marginal\_normality*. If true, the
    marginals are transformed to marginal standard normality according
    to Otneim and Tjøstheim (2017). This is the **default** method.

See the documentation of the *lg\_main()*-function for further details.
We can now construct the lg-object using the default configuration by
running

``` r
library(lg)
lg_object <- lg_main(x)
```

# Estimation of density functions

We can then specify a set of grid points and estimate the probability
density function of *x* using the *dlg()*-function. We choose a set of
grid ponts that go diagonally through R^4, estimate, and plot the result
as
follows:

``` r
grid <- matrix(rep(seq(-.03, .03, length.out = 100), 4), ncol = 4, byrow = FALSE)
density_estimate <- dlg(lg_object = lg_object, grid = grid)
# plot(grid[,1], density_estimate$f_est, type = "l",
#     xlab = "Diagonal grid point", ylab = "Estimated density")
```

# Estimation of conditional densities

If we want to calculate conditional density functions, we must take care
to notice the *order* of the columns in our data set. This is because
the estimation routine, implemented in the *clg()*-function, will always
assume that the independent variables come first. Looking at the top of
our data set:

``` r
head(x)
#>               DAX          SMI          CAC         FTSE
#> [1,] -0.009326550  0.006178360 -0.012658756  0.006770286
#> [2,] -0.004422175 -0.005880448 -0.018740638 -0.004889587
#> [3,]  0.009003794  0.003271184 -0.005779182  0.009027020
#> [4,] -0.001778217  0.001483372  0.008743353  0.005771847
#> [5,] -0.004676712 -0.008933417 -0.005120160 -0.007230164
#> [6,]  0.012427042  0.006737244  0.011714353  0.008517217
```

we see that DAX comes first. Say that we want to estimate the
conditional density of DAX, given that SMI = CAC = FTSE = 0. We do that
by
running

``` r
grid <- matrix(seq(-.03, .03, length.out = 100), ncol = 1)   # The grid must be a matrix
condition <- c(0, 0, 0)                                      # Value of dependent variables
cond_dens_est <- clg(lg_object = lg_object, 
                     grid = grid,
                     condition = condition)
# plot(grid, cond_dens_est$f_est, type = "l",
#     xlab = "DAX", ylab = "Estimated conditional density")
```

If we want to estimate the conditional density of CAC and FTSE given DAX
and SMI, for example, we must first shuffle the data so that CAC and
FTSE come first, and supply the conditional value for DAX and SMI
through the vector *condition*, now having two elements.

# Independence tests and test of financial contagion

The following statistical tests are available:

  - Test for independence between two stochastic vectors by Berentsen
    and Tjøstheim (2014).
  - Test for serial dependence in a time series by Lacal and Tjøstheim
    (2017a).
  - Test for cross-dependence between two time series by Lacal and
    Tjøstheim (2017b).
  - Test for financial contagion during crises by Støve et al. (2014).

Let us quickly demonstrate their implementation. For the first test, we
generate some data from a bivariate \(t\)-distribution. They are
uncorrelated, but not independent. In order to test for independence, we
create an lg-object, and apply the function *ind\_test()*. One may of
course change the estimation method, the bandwidths and so in the call
to *lg\_main()*, and there are furher options available in
*ind\_test()*. For this illustration we use only 20 bootstrap
replications, but this must of course be significantly higher in
practical applications.

``` r
set.seed(1)
x <- mvtnorm::rmvt(n = 100, df = 2)
lg_object <- lg_main(x, est_method = "5par")
test_result <- ind_test(lg_object, n_rep = 20)
test_result$p_value
#> [1] 0
```

The test for serial dependence in a time series X(t) can be performed in
exactly the same way by collecting *X(t)* and *X(t-k)* as columns in the
data set, for some lag *k*. For a test for serial cross dependence
between *X(t)* and *Y(t)* one must collect *X(t)* and *Y(t-k)* as
columns in the data set, but also choose either `bootstrap_type =
"block"` or `bootstrap_type = "stationary"` in order to correctly
resample under the null hypothesis. We refer to the original article
Lacal and Tjøstheim (2017b) for details on this.

In order to perform the test for financial contagion, one must collect
the non-crisis and crisis data in separate data sets, and create
separate lg-objects, and then apply the *cont\_test()*-function.

# References

Berentsen, Geir Drage, Tore Selland Kleppe, and Dag Tjøstheim.
“Introducing localgauss, an R package for estimating and visualizing
local Gaussian correlation.” Journal of Statistical Software 56.1
(2014): 1-18.

Berentsen, Geir Drage, and Dag Tjøstheim. “Recognizing and visualizing
departures from independence in bivariate data using local Gaussian
correlation.” Statistics and Computing 24.5 (2014): 785-801.

Lacal, Virginia, and Dag Tjøstheim. “Local Gaussian autocorrelation and
tests for serial independence.” Journal of Time Series Analysis 38.1
(2017a): 51-71.

Lacal, Virginia, and Dag Tjøstheim. “Estimating and testing nonlinear
local dependence between two time series.” Journal of Business &
Economic Statistics just-accepted (2017b).

Otneim, Håkon, and Dag Tjøstheim. “The locally gaussian density
estimator for multivariate data.” Statistics and Computing 27.6 (2017):
1595-1616.

Otneim, Håkon, and Dag Tjøstheim. “Conditional density estimation using
the local Gaussian correlation.” Statistics and Computing 28.2 (2018):
303-321.

Støve, Bård, Dag Tjøstheim, and Karl Ove Hufthammer. “Using local
Gaussian correlation in a nonlinear re-examination of financial
contagion.” Journal of Empirical Finance 25 (2014): 62-82.

Tjøstheim, Dag, & Hufthammer, Karl Ove (2013). Local Gaussian
correlation: a new measure of dependence. Journal of Econometrics,
172(1), 33-48.


NEWS.md
# lg 0.4.1

   * Fixed small bug causing error in r-devel

# lg 0.4.0

   * Added bootstrap confidence intervals as option in dlg.
   * Added the estimation method "trivariate" for joint estimation and independence testing 
     with trivariate data.

# lg 0.3.0
 
   * Added contagion test and independence tests.
   * Added bw_simple for quick construction of bandwidth object.
   * Switches to old logspline algorithm if the new fails (very rare). 
   * Better NA handling, especially in the test for conditional independence

# lg 0.2.0 

   * Added function partial_cor() for calculating the local partial correlations
     from a clg-object.
   * Added function ci_test() for conducting a test for conditional 
     independence.
   * Added functionality for normalizing estimated (conditional) densities using
     a MC approximation.
   * Local (partial) correlations now comes with approximate confidence 
     intervals.

# lg 0.1.0 

   * Initial release


R/main_function.R

# The main function that creates the 'lg'-object
# ----------------------------------------------

#'Create an \code{lg} object
#'
#'Create an \code{lg}-object, that can be used to estimate local Gaussian
#'correlations, unconditional and conditional densities, local partial
#'correlation and for testing purposes.
#'
#'This is the main function in the package. It lets the user supply a data set
#'and set a number of options, which is then used to prepare an \code{lg} object
#'that can be supplied to other functions in the package, such as \code{dlg}
#'(density estimation), \code{clg} (conditional density estimation). The details
#'has been laid out in Otneim & Tjøstheim (2017) and Otneim & Tjøstheim (2018).
#'
#'The papers mentioned above deal with the estimation of multivariate density
#'functions and conditional density functions. The idea is to fit a multivariate
#'Normal locally to the unknown density function by first transforming the data
#'to marginal standard normality, and then estimate the local correlations
#'\strong{pairwise}. The local means and local standard deviations are held
#'fixed and constantly equal to 0 and 1 respectively to reflect the knowledge
#'that the marginals are approximately standard normal. Use \code{est_method =
#'"1par"} for this strategy, which means that we only estimate one local
#'parameter (the correlation) for each pair, and note that this method requires
#'marginally standard normal data. If \code{est_method = "1par"} and
#'\code{transform_to_marginal_normality = FALSE} the function will throw a
#'warning. It might be okay though, if you know that the data are marginally
#'standard normal already.
#'
#'The second option is \code{est_method = "5par_marginals_fixed"} which is more
#'flexible than \code{"1par"}. This method will estimate univariate local
#'Gaussian fits to each marginal, thus producing local estimates of the local
#'means: \eqn{\mu_i(x_i)} and \eqn{\sigma_i(x_i)} that will be held fixed in the
#'next step when the \strong{pairwise} local correlations are estimated. This
#'method can in many situations provide a better fit, even if the marginals are
#'standard normal. It also opens up for creating a multivariate locally Gaussian
#'fit to any density without having to transform the marginals if you for some
#'reason want to avoid that.
#'
#'The third option is \code{est_method = "5par"}, which is a full nonparametric
#'locally Gaussian fit of a bivariate density as laid out and used by Tjøstheim
#'& Hufthammer (2013) and others. This is simply a wrapper for the
#'\code{localgauss}-package by Berentsen et.al. (2014).
#'
#'A recent option is described by Otneim and Tjøstheim (2019), who allow a full
#'trivariate fit to a three dimensional data set that is transformed to marginal
#'standard normality in the context of their test for conditional independence
#'(see \code{?ci_test} for details), but this can of course be used as an option
#'to estimate three-variate density functions as well.
#'
#'@param x A matrix or data frame with data, on column per variable, one row per
#'  observation.
#'@param bw_method The method used for bandwidth selection. Must be either
#'  \code{"cv"} (cross-validation, slow, but accurate) or \code{"plugin"} (fast,
#'  but crude).
#'@param est_method The estimation method, must be either "1par", "5par",
#'  "5par_marginals_fixed" or "trivariate". (see details).
#'@param transform_to_marginal_normality Logical, \code{TRUE} if we want to
#'  transform our data to marginal standard normality. This is assumed by method
#'  "1par", but can of course be skipped using this argument if it has been done
#'  already.
#'@param bw Bandwidth object if it has already been calculated.
#'@param plugin_constant_marginal The constant \code{c} in \code{cn^a} used for
#'  finding the plugin bandwidth for locally Gaussian marginal density
#'  estimates, which we need if estimation method is "5par_marginals_fixed".
#'@param plugin_exponent_marginal The constant \code{a} in \code{cn^a} used for
#'  finding the plugin bandwidth for locally Gaussian marginal density
#'  estimates, which we need if estimation method is "5par_marginals_fixed".
#'@param plugin_constant_joint The constant \code{c} in \code{cn^a} used for
#'  finding the plugin bandwidth for estimating the pairwise local Gaussian
#'  correlation between two variables.
#'@param plugin_exponent_joint The constant \code{a} in \code{cn^a} used for
#'  finding the plugin bandwidth for estimating the pairwise local Gaussian
#'  correlation between two variables.
#'@param tol_marginal The absolute tolerance in the optimization for finding the
#'  marginal bandwidths, passed on to the \code{optim}-function.
#'@param tol_joint The absolute tolerance in the optimization for finding the
#'  joint bandwidths. Passed on to the \code{optim}-function.
#'
#' @examples
#'   x <- cbind(rnorm(100), rnorm(100), rnorm(100))
#'
#'   # Quick example
#'   lg_object1 <- lg_main(x, bw_method = "plugin", est_method = "1par")
#'
#'   # In the simulation experiments in Otneim & Tjøstheim (2017a),
#'   # the cross-validation bandwidth selection is used:
#'   \dontrun{
#'   lg_object2 <- lg_main(x, bw_method = "cv", est_method = "1par")
#'   }
#'
#'   # If you do not wish to transform the data to standard normality,
#'   # use the five parameter fit:
#'   lg_object3 <- lg_main(x, est_method = "5par_marginals_fixed",
#'                   transform_to_marginal_normality = FALSE)
#'
#'   # In the bivariate case, you can use the full nonparametric fit:
#'   x_biv <- cbind(rnorm(100), rnorm(100))
#'   lg_object4 <- lg_main(x_biv, est_method = "5par",
#'                   transform_to_marginal_normality = FALSE)
#'
#'   # Whichever method you choose, the lg-object can now be passed on
#'   # to the dlg- or clg-functions for evaluation of the density or
#'   # conditional density estimate. Control the grid with the grid
#'   # argument.
#'   grid1 <- x[1:10,]
#'   dens_est <- dlg(lg_object1, grid = grid1)
#'
#'   # The conditional density of X1 given X2 = 1 and X2 = 0:
#'   grid2 <- matrix(-3:3, ncol = 1)
#'   c_dens_est <- clg(lg_object1, grid = grid2, condition = c(1, 0))
#'
#'@references
#'
#'Berentsen, Geir Drage, Tore Selland Kleppe, and Dag Tjøstheim. "Introducing
#'localgauss, an R package for estimating and visualizing local Gaussian
#'correlation." Journal of Statistical Software 56.1 (2014): 1-18.
#'
#'Hufthammer, Karl Ove, and Dag Tjøstheim. "Local Gaussian Likelihood and Local
#'Gaussian Correlation" PhD Thesis of Karl Ove Hufthammer, University of Bergen,
#'2009.
#'
#'Otneim, Håkon, and Dag Tjøstheim. "The locally gaussian density estimator for
#'multivariate data." Statistics and Computing 27, no. 6 (2017): 1595-1616.
#'
#'Otneim, Håkon, and Dag Tjøstheim. "Conditional density estimation using
#'the local Gaussian correlation" Statistics and Computing 28, no. 2 (2018):
#'303-321.
#'
#'Otneim, Håkon, and Dag Tjøstheim. "The local Gaussian partial correlation"
#'Working paper (2019).
#'
#'Tjøstheim, D., & Hufthammer, K. O. (2013). Local Gaussian correlation: a new
#'measure of dependence. Journal of Econometrics, 172(1), 33-48.
#'
#'@export
lg_main <- function(x,
               bw_method = "plugin",
               est_method = "1par",
               transform_to_marginal_normality = TRUE,
               bw = NULL,
               plugin_constant_marginal = 1.75,
               plugin_constant_joint = 1.75,
               plugin_exponent_marginal = -1/5,
               plugin_exponent_joint = -1/6,
               tol_marginal = 10^(-3),
               tol_joint = 10^(-3)) {

    # Sanity checks
    if(is.atomic(x) & is.vector(x)) x <- matrix(x, ncol = 1)
    x <- check_data(x, type = "data")
    check_est_method(est_method)
    if((est_method == "5par") & (ncol(x) != 2)) {
        stop("Data must be bivariate if estimation method is '5par'")
    }
    if((est_method == "1par") & (transform_to_marginal_normality == FALSE)) {
        warning("Estimation method '1par' assumes marginal standard normality.")
    }
    if((est_method == "trivariate") & (transform_to_marginal_normality == FALSE)) {
        warning("Estimation method 'trivariate' assumes marginal standard normality.")
    }
    if((est_method == "trivariate") & (ncol(x) != 3)) {
        stop("Data must be trivariate if estimation method is 'trivariate'")
    }

    # Return a list
    ret <- list()
    ret$x <- x
    ret$bw_method <- bw_method
    ret$est_method <- est_method
    ret$transform_to_marginal_normality <- transform_to_marginal_normality

    # Transformation
    if(transform_to_marginal_normality) {
        transformed <- trans_normal(x = x)
        ret$transformed_data <- transformed$transformed_data
        ret$trans_new <- transformed$trans_new
    } else {
        ret$transformed_data <- x
        ret$trans_new <- NA
    }

    # Bandwidth selection
    if(is.null(bw)) {
        bw <- bw_select(ret$transformed_data,
                        bw_method = bw_method,
                        est_method = est_method,
                        plugin_constant_marginal = plugin_constant_marginal,
                        plugin_exponent_marginal =  plugin_exponent_marginal,
                        plugin_constant_joint = plugin_constant_joint,
                        plugin_exponent_joint = plugin_exponent_joint,
                        tol_marginal = tol_marginal,
                        tol_joint = tol_joint)
    }
    ret$bw <- bw

    # Add the rest of the information to the object that we return
    ret$plugin_constant_marginal <- plugin_constant_marginal
    ret$plugin_constant_joint    <- plugin_constant_joint
    ret$plugin_exponent_marginal <- plugin_exponent_marginal
    ret$plugin_exponent_joint    <- plugin_exponent_joint
    ret$tol_marginal             <- tol_marginal
    ret$tol_joint                <- tol_joint

    class(ret) <- "lg"

    return(ret)
}
